// bitski/lib/auth/user-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../constants");
const localstorage_store_1 = require("../utils/localstorage-store");
const user_1 = require("./user");
class UserStore {
    constructor(clientId, store) {
        this.clientId = clientId;
        this.store = store || new localstorage_store_1.LocalStorageStore();
        this.user = this.fetchUser();
    }
    get currentUser() {
        return this.user || this.fetchUser();
    }
    get storageKey() {
        return `${constants_1.USER_KEY}.${this.clientId}`;
    }
    set(user) {
        this.user = user;
        this.cacheUser(user);
    }
    clear() {
        this.user = undefined;
        this.cacheUser(undefined);
    }
    fetchUser() {
        const userData = this.store.getItem(this.storageKey);
        if (userData) {
            return user_1.User.fromString(userData);
        }
    }
    cacheUser(user) {
        if (user) {
            this.store.setItem(this.storageKey, user.toStorageString());
        }
        else {
            this.store.clearItem(this.storageKey);
        }
    }
}
exports.UserStore = UserStore;
