// bitski/lib/auth/user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class User {
    constructor(id, accounts, email, emailVerified, phone, phoneNumberVerified) {
        this.id = id;
        this.accounts = accounts;
        this.email = email;
        this.emailVerified = emailVerified;
        this.phoneNumber = phone;
        this.phoneNumberVerified = phoneNumberVerified;
    }
    static fromJson(json) {
        return new User(json.sub, json.accounts, json.email, json.email_verified, json.phone_number, json.phone_number_verified);
    }
    static fromString(s) {
        let parsed;
        try {
            parsed = JSON.parse(s);
        }
        catch (e) {
            return;
        }
        if (parsed.id) {
            return new User(parsed.id, parsed.accounts, parsed.email, parsed.emailVerified, parsed.phoneNumber, parsed.phoneNumberVerified);
        }
        return;
    }
    toStorageString() {
        return JSON.stringify({
            accounts: this.accounts,
            email: this.email,
            emailVerified: this.emailVerified,
            id: this.id,
            phoneNumber: this.phoneNumber,
            phoneNumberVerified: this.phoneNumberVerified,
        });
    }
}
exports.User = User;
