// bitski/dist/esm/auth/user-store.js
import { USER_KEY } from '../constants';
import { LocalStorageStore } from '../utils/localstorage-store';
import { User } from './user';
export class UserStore {
    constructor(clientId, store) {
        this.clientId = clientId;
        this.store = store || new LocalStorageStore();
        this.user = this.fetchUser();
    }
    get currentUser() {
        return this.user || this.fetchUser();
    }
    get storageKey() {
        return `${USER_KEY}.${this.clientId}`;
    }
    set(user) {
        this.user = user;
        this.cacheUser(user);
    }
    clear() {
        this.user = undefined;
        this.cacheUser(undefined);
    }
    fetchUser() {
        const userData = this.store.getItem(this.storageKey);
        if (userData) {
            return User.fromString(userData);
        }
    }
    cacheUser(user) {
        if (user) {
            this.store.setItem(this.storageKey, user.toStorageString());
        }
        else {
            this.store.clearItem(this.storageKey);
        }
    }
}
