// @0xcert/ethereum-asset-ledger/dist/core/ledger.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const ethereum_generic_provider_1 = require("@0xcert/ethereum-generic-provider");
const __1 = require("..");
const approve_account_1 = require("../mutations/approve-account");
const create_asset_1 = require("../mutations/create-asset");
const deploy_1 = require("../mutations/deploy");
const destroy_asset_1 = require("../mutations/destroy-asset");
const grant_abilities_1 = require("../mutations/grant-abilities");
const revoke_abilities_1 = require("../mutations/revoke-abilities");
const revoke_asset_1 = require("../mutations/revoke-asset");
const safe_transfer_1 = require("../mutations/safe-transfer");
const set_abilities_1 = require("../mutations/set-abilities");
const set_approval_for_all_1 = require("../mutations/set-approval-for-all");
const set_enabled_1 = require("../mutations/set-enabled");
const transfer_1 = require("../mutations/transfer");
const update_1 = require("../mutations/update");
const update_asset_1 = require("../mutations/update-asset");
const get_abilities_1 = require("../queries/get-abilities");
const get_account_asset_id_at_1 = require("../queries/get-account-asset-id-at");
const get_approved_account_1 = require("../queries/get-approved-account");
const get_asset_1 = require("../queries/get-asset");
const get_asset_account_1 = require("../queries/get-asset-account");
const get_asset_id_at_1 = require("../queries/get-asset-id-at");
const get_balance_1 = require("../queries/get-balance");
const get_capabilities_1 = require("../queries/get-capabilities");
const get_info_1 = require("../queries/get-info");
const is_approved_for_all_1 = require("../queries/is-approved-for-all");
const is_enabled_1 = require("../queries/is-enabled");
class AssetLedger {
    constructor(provider, id) {
        this._provider = provider;
        this._id = this._provider.encoder.normalizeAddress(id);
    }
    static deploy(provider, recipe) {
        return __awaiter(this, void 0, void 0, function* () {
            return deploy_1.default(provider, recipe);
        });
    }
    static getInstance(provider, id) {
        return new this(provider, id);
    }
    get id() {
        return this._id;
    }
    get provider() {
        return this._provider;
    }
    getAbilities(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return get_abilities_1.default(this, accountId);
        });
    }
    getApprovedAccount(assetId) {
        return __awaiter(this, void 0, void 0, function* () {
            return get_approved_account_1.default(this, assetId);
        });
    }
    getAssetAccount(assetId) {
        return __awaiter(this, void 0, void 0, function* () {
            return get_asset_account_1.default(this, assetId);
        });
    }
    getAsset(assetId) {
        return __awaiter(this, void 0, void 0, function* () {
            return get_asset_1.default(this, assetId);
        });
    }
    getBalance(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return get_balance_1.default(this, accountId);
        });
    }
    getCapabilities() {
        return __awaiter(this, void 0, void 0, function* () {
            return get_capabilities_1.default(this);
        });
    }
    getInfo() {
        return __awaiter(this, void 0, void 0, function* () {
            return get_info_1.default(this);
        });
    }
    getAssetIdAt(index) {
        return __awaiter(this, void 0, void 0, function* () {
            return get_asset_id_at_1.default(this, index);
        });
    }
    getAccountAssetIdAt(accountId, index) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return get_account_asset_id_at_1.default(this, accountId, index);
        });
    }
    isApprovedAccount(assetId, accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return accountId === (yield get_approved_account_1.default(this, assetId));
        });
    }
    isTransferable() {
        return __awaiter(this, void 0, void 0, function* () {
            return is_enabled_1.default(this);
        });
    }
    approveAccount(assetId, accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return approve_account_1.default(this, accountId, assetId);
        });
    }
    disapproveAccount(assetId) {
        return __awaiter(this, void 0, void 0, function* () {
            return approve_account_1.default(this, '0x0000000000000000000000000000000000000000', assetId);
        });
    }
    grantAbilities(accountId, abilities) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return grant_abilities_1.default(this, accountId, __1.getBitfieldFromAbilities(abilities));
        });
    }
    setAbilities(accountId, abilities) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return set_abilities_1.default(this, accountId, __1.getBitfieldFromAbilities(abilities));
        });
    }
    createAsset(recipe) {
        return __awaiter(this, void 0, void 0, function* () {
            const receiverId = this._provider.encoder.normalizeAddress(recipe.receiverId);
            return create_asset_1.default(this, receiverId, recipe.id, `0x${recipe.imprint}`);
        });
    }
    destroyAsset(assetId) {
        return __awaiter(this, void 0, void 0, function* () {
            return destroy_asset_1.default(this, assetId);
        });
    }
    revokeAbilities(accountId, abilities) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return revoke_abilities_1.default(this, accountId, __1.getBitfieldFromAbilities(abilities));
        });
    }
    revokeAsset(assetId) {
        return __awaiter(this, void 0, void 0, function* () {
            return revoke_asset_1.default(this, assetId);
        });
    }
    transferAsset(recipe) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!recipe.senderId) {
                recipe.senderId = this.provider.accountId;
            }
            const senderId = this._provider.encoder.normalizeAddress(recipe.senderId);
            const receiverId = this._provider.encoder.normalizeAddress(recipe.receiverId);
            return this.provider.unsafeRecipientIds.indexOf(recipe.receiverId) !== -1
                ? transfer_1.default(this, senderId, receiverId, recipe.id)
                : safe_transfer_1.default(this, senderId, receiverId, recipe.id, recipe.data);
        });
    }
    enableTransfers() {
        return __awaiter(this, void 0, void 0, function* () {
            return set_enabled_1.default(this, true);
        });
    }
    disableTransfers() {
        return __awaiter(this, void 0, void 0, function* () {
            return set_enabled_1.default(this, false);
        });
    }
    updateAsset(assetId, recipe) {
        return __awaiter(this, void 0, void 0, function* () {
            return update_asset_1.default(this, assetId, `0x${recipe.imprint}`);
        });
    }
    update(recipe) {
        return __awaiter(this, void 0, void 0, function* () {
            return update_1.default(this, recipe.uriPrefix, recipe.uriPostfix);
        });
    }
    approveOperator(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return set_approval_for_all_1.default(this, accountId, true);
        });
    }
    disapproveOperator(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            return set_approval_for_all_1.default(this, accountId, false);
        });
    }
    isApprovedOperator(accountId, operatorId) {
        return __awaiter(this, void 0, void 0, function* () {
            accountId = this._provider.encoder.normalizeAddress(accountId);
            operatorId = this._provider.encoder.normalizeAddress(operatorId);
            return is_approved_for_all_1.default(this, accountId, operatorId);
        });
    }
    getContext() {
        return [
            {
                name: 'Transfer',
                topic: '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
                types: [
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'from',
                        type: 'address',
                    },
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'to',
                        type: 'address',
                    },
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'tokenId',
                        type: 'uint256',
                    },
                ],
            },
            {
                name: 'Approval',
                topic: '0x8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925',
                types: [
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'owner',
                        type: 'address',
                    },
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'approved',
                        type: 'address',
                    },
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'tokenId',
                        type: 'uint256',
                    },
                ],
            },
            {
                name: 'ApprovalForAll',
                topic: '0x17307eab39ab6107e8899845ad3d59bd9653f200f220920489ca2b5937696c31',
                types: [
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'owner',
                        type: 'address',
                    },
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'operator',
                        type: 'address',
                    },
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.NORMAL,
                        name: 'approved',
                        type: 'bool',
                    },
                ],
            },
            {
                name: 'IsPaused',
                topic: '0xff4a5dbbab6b1963d10f5edd139f33a7987ecb3c4f65969be77ddba28d946594',
                types: [
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.NORMAL,
                        name: 'isPaused',
                        type: 'bool',
                    },
                ],
            },
            {
                name: 'TokenImprintUpdate',
                topic: '0xda81dc3bdaf65b621535dbd08fc4032c7e22d5ccab12c8aa2ad010611471d627',
                types: [
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'tokenId',
                        type: 'uint256',
                    },
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.NORMAL,
                        name: 'imprint',
                        type: 'bytes32',
                    },
                ],
            },
            {
                name: 'SetAbilities',
                topic: '0xd1d59d2d212a435434e7a4a4676427610dfe2b6268b01e541d280d65bf3d6b90',
                types: [
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'target',
                        type: 'address',
                    },
                    {
                        kind: ethereum_generic_provider_1.MutationEventTypeKind.INDEXED,
                        name: 'abilities',
                        type: 'uint256',
                    },
                ],
            },
        ];
    }
}
exports.AssetLedger = AssetLedger;
