// @0xcert/ethereum-asset-ledger/dist/mutations/set-enabled.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const ethereum_generic_provider_1 = require("@0xcert/ethereum-generic-provider");
const functionSignature = '0xbedb86fb';
const inputTypes = ['bool'];
function default_1(ledger, enabled) {
    return __awaiter(this, void 0, void 0, function* () {
        const attrs = {
            from: ledger.provider.accountId,
            to: ledger.id,
            data: functionSignature + ledger.provider.encoder.encodeParameters(inputTypes, [!enabled]).substr(2),
        };
        const res = yield ledger.provider.post({
            method: 'eth_sendTransaction',
            params: [attrs],
        });
        return new ethereum_generic_provider_1.Mutation(ledger.provider, res.result, ledger);
    });
}
exports.default = default_1;
