// bitski-provider/lib/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./bitski-engine", "./auth/access-token", "./subproviders/authenticated-fetch", "./network", "./errors/server-error", "./errors/provider-error"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var bitski_engine_1 = require("./bitski-engine");
    exports.BitskiEngine = bitski_engine_1.BitskiEngine;
    var access_token_1 = require("./auth/access-token");
    exports.AccessToken = access_token_1.AccessToken;
    var authenticated_fetch_1 = require("./subproviders/authenticated-fetch");
    exports.AuthenticatedFetchSubprovider = authenticated_fetch_1.AuthenticatedFetchSubprovider;
    var network_1 = require("./network");
    exports.Mainnet = network_1.Mainnet;
    exports.Rinkeby = network_1.Rinkeby;
    exports.Kovan = network_1.Kovan;
    var server_error_1 = require("./errors/server-error");
    exports.ServerError = server_error_1.ServerError;
    var provider_error_1 = require("./errors/provider-error");
    exports.ProviderError = provider_error_1.ProviderError;
    exports.ProviderErrorCode = provider_error_1.ProviderErrorCode;
});
