// @0xcert/ethereum-generic-provider/dist/core/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SignMethod;
(function (SignMethod) {
    SignMethod[SignMethod["ETH_SIGN"] = 0] = "ETH_SIGN";
    SignMethod[SignMethod["TREZOR"] = 1] = "TREZOR";
    SignMethod[SignMethod["NO_PREFIX"] = 2] = "NO_PREFIX";
    SignMethod[SignMethod["PERSONAL_SIGN"] = 3] = "PERSONAL_SIGN";
})(SignMethod = exports.SignMethod || (exports.SignMethod = {}));
var MutationEventTypeKind;
(function (MutationEventTypeKind) {
    MutationEventTypeKind[MutationEventTypeKind["NORMAL"] = 0] = "NORMAL";
    MutationEventTypeKind[MutationEventTypeKind["INDEXED"] = 1] = "INDEXED";
})(MutationEventTypeKind = exports.MutationEventTypeKind || (exports.MutationEventTypeKind = {}));
var NetworkKind;
(function (NetworkKind) {
    NetworkKind[NetworkKind["LIVE"] = 0] = "LIVE";
    NetworkKind[NetworkKind["ROPSTEN"] = 1] = "ROPSTEN";
    NetworkKind[NetworkKind["RINKEBY"] = 2] = "RINKEBY";
})(NetworkKind = exports.NetworkKind || (exports.NetworkKind = {}));
