// @0xcert/scaffold/dist/core/gateway.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ActionsOrderActionKind;
(function (ActionsOrderActionKind) {
    ActionsOrderActionKind[ActionsOrderActionKind["CREATE_ASSET"] = 1] = "CREATE_ASSET";
    ActionsOrderActionKind[ActionsOrderActionKind["TRANSFER_ASSET"] = 2] = "TRANSFER_ASSET";
    ActionsOrderActionKind[ActionsOrderActionKind["TRANSFER_VALUE"] = 3] = "TRANSFER_VALUE";
    ActionsOrderActionKind[ActionsOrderActionKind["UPDATE_ASSET_IMPRINT"] = 4] = "UPDATE_ASSET_IMPRINT";
    ActionsOrderActionKind[ActionsOrderActionKind["SET_ABILITIES"] = 5] = "SET_ABILITIES";
    ActionsOrderActionKind[ActionsOrderActionKind["DESTROY_ASSET"] = 6] = "DESTROY_ASSET";
})(ActionsOrderActionKind = exports.ActionsOrderActionKind || (exports.ActionsOrderActionKind = {}));
var OrderKind;
(function (OrderKind) {
    OrderKind[OrderKind["ASSET_LEDGER_DEPLOY_ORDER"] = 1] = "ASSET_LEDGER_DEPLOY_ORDER";
    OrderKind[OrderKind["VALUE_LEDGER_DEPLOY_ORDER"] = 2] = "VALUE_LEDGER_DEPLOY_ORDER";
    OrderKind[OrderKind["FIXED_ACTIONS_ORDER"] = 3] = "FIXED_ACTIONS_ORDER";
    OrderKind[OrderKind["DYNAMIC_ACTIONS_ORDER"] = 4] = "DYNAMIC_ACTIONS_ORDER";
    OrderKind[OrderKind["SIGNED_FIXED_ACTIONS_ORDER"] = 5] = "SIGNED_FIXED_ACTIONS_ORDER";
    OrderKind[OrderKind["SIGNED_DYNAMIC_ACTIONS_ORDER"] = 6] = "SIGNED_DYNAMIC_ACTIONS_ORDER";
    OrderKind[OrderKind["ASSET_SET_OPERATOR_ORDER"] = 7] = "ASSET_SET_OPERATOR_ORDER";
    OrderKind[OrderKind["DAPP_VALUE_APPROVE_ORDER"] = 8] = "DAPP_VALUE_APPROVE_ORDER";
})(OrderKind = exports.OrderKind || (exports.OrderKind = {}));
class DappValueApproveOrder {
}
exports.DappValueApproveOrder = DappValueApproveOrder;
class AssetSetOperatorOrder {
}
exports.AssetSetOperatorOrder = AssetSetOperatorOrder;
class DynamicActionsOrder {
}
exports.DynamicActionsOrder = DynamicActionsOrder;
class SignedDynamicActionsOrder {
}
exports.SignedDynamicActionsOrder = SignedDynamicActionsOrder;
class FixedActionsOrder {
}
exports.FixedActionsOrder = FixedActionsOrder;
class SignedFixedActionsOrder {
}
exports.SignedFixedActionsOrder = SignedFixedActionsOrder;
class AssetLedgerDeployOrder {
}
exports.AssetLedgerDeployOrder = AssetLedgerDeployOrder;
class ValueLedgerDeployOrder {
}
exports.ValueLedgerDeployOrder = ValueLedgerDeployOrder;
