// @0xcert/scaffold/dist/core/issues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ProviderIssue;
(function (ProviderIssue) {
    ProviderIssue["GENERAL"] = "0";
    ProviderIssue["CONTRACT_ERROR"] = "1";
    ProviderIssue["ACTIONS_ORDER_KIND_NOT_SUPPORTED"] = "5000001";
    ProviderIssue["DYNAMIC_ACTIONS_ORDER_SIGNATURES"] = "5000002";
    ProviderIssue["FIXED_ACTIONS_ORDER_SIGNATURES"] = "5000003";
    ProviderIssue["SIGNED_DYNAMIC_ACTIONS_ORDER_SIGNATURES"] = "5000004";
    ProviderIssue["SIGNED_FIXED_ACTIONS_ORDER_SIGNATURES"] = "5000005";
    ProviderIssue["NO_RECEIVER_ID"] = "5000006";
    ProviderIssue["PROXY_KIND_NOT_SUPPORTED"] = "5000007";
    ProviderIssue["SENDER_ID_NOT_A_SIGNER"] = "5000008";
    ProviderIssue["SENDER_ID_AND_RECEIVER_ID_MISSING"] = "5000009";
    ProviderIssue["ACTION_KIND_NOT_SUPPORTED"] = "5000010";
    ProviderIssue["ERC20_APPROVAL_RACE_CONDITION"] = "5000011";
    ProviderIssue["ERC20_NOT_ENOUGH_BALANCE"] = "001001";
    ProviderIssue["ERC20_NOT_ENOUGH_ALLOWANCE"] = "001002";
    ProviderIssue["ERC721_ZERO_ADDRESS"] = "003001";
    ProviderIssue["ERC721_NOT_VALID_NFT"] = "003002";
    ProviderIssue["ERC721_NOT_OWNER_OR_OPERATOR"] = "003003";
    ProviderIssue["ERC721_NOT_OWNER_APPROWED_OR_OPERATOR"] = "003004";
    ProviderIssue["ERC721_NOT_ABLE_TO_RECEIVE_NFT"] = "003005";
    ProviderIssue["ERC721_NFT_ALREADY_EXISTS"] = "003006";
    ProviderIssue["ERC721_METADATA_ZERO_ADDRESS"] = "004001";
    ProviderIssue["ERC721_METADATA_NOT_VALID_NFT"] = "004002";
    ProviderIssue["ERC721_METADATA_NOT_OWNER_OR_OPERATOR"] = "004003";
    ProviderIssue["ERC721_METADATA_NOT_OWNER_APPROWED_OR_OPERATOR"] = "004004";
    ProviderIssue["ERC721_METADATA_NOT_ABLE_TO_RECEIVE_NFT"] = "004005";
    ProviderIssue["ERC721_METADATA_NFT_ALREADY_EXISTS"] = "004006";
    ProviderIssue["ERC721_ENUMERABLE_ZERO_ADDRESS"] = "005001";
    ProviderIssue["ERC721_ENUMERABLE_NOT_VALID_NFT"] = "005002";
    ProviderIssue["ERC721_ENUMERABLE_NOT_OWNER_OR_OPERATOR"] = "005003";
    ProviderIssue["ERC721_ENUMERABLE_NOT_OWNER_APPROWED_OR_OPERATOR"] = "005004";
    ProviderIssue["ERC721_ENUMERABLE_NOT_ABLE_TO_RECEIVE_NFT"] = "005005";
    ProviderIssue["ERC721_ENUMERABLE_NFT_ALREADY_EXISTS"] = "005006";
    ProviderIssue["ERC721_ENUMERABLE_INVALID_INDEX"] = "005007";
    ProviderIssue["ERC721_METADATA_ENUMERABLE_ZERO_ADDRESS"] = "006001";
    ProviderIssue["ERC721_METADATA_ENUMERABLE_NOT_VALID_NFT"] = "006002";
    ProviderIssue["ERC721_METADATA_ENUMERABLE_NOT_OWNER_OR_OPERATOR"] = "006003";
    ProviderIssue["ERC721_METADATA_ENUMERABLE_NOT_OWNER_APPROWED_OR_OPERATOR"] = "006004";
    ProviderIssue["ERC721_METADATA_ENUMERABLE_NOT_ABLE_TO_RECEIVE_NFT"] = "006005";
    ProviderIssue["ERC721_METADATA_ENUMERABLE_NFT_ALREADY_EXISTS"] = "006006";
    ProviderIssue["ERC721_METADATA_ENUMERABLE_INVALID_INDEX"] = "006007";
    ProviderIssue["XCERT_CAPABILITY_NOT_SUPPORTED"] = "007001";
    ProviderIssue["XCERT_TRANSFERS_DISABLED"] = "007002";
    ProviderIssue["XCERT_NOT_VALID_XCERT"] = "007003";
    ProviderIssue["XCERT_NOT_OWNER_OR_OPERATOR"] = "007004";
    ProviderIssue["XCERT_INVALID_SIGNATURE"] = "007005";
    ProviderIssue["XCERT_INVALID_SIGNATURE_KIND"] = "007006";
    ProviderIssue["XCERT_CLAIM_PERFORMED"] = "007007";
    ProviderIssue["XCERT_CLAIM_EXPIRED"] = "007008";
    ProviderIssue["XCERT_CLAIM_CANCELED"] = "007009";
    ProviderIssue["XCERT_NOT_OWNER"] = "007010";
    ProviderIssue["SAFE_MATH_OVERFLOW"] = "008001";
    ProviderIssue["SAFE_MATH_SUBTRAHEND_GREATER_THEN_MINUEND"] = "008002";
    ProviderIssue["SAFE_MATH_DIVISION_BY_ZERO"] = "008003";
    ProviderIssue["XCERT_DEPLOY_GATEWAY_INVALID_SIGNATURE_KIND"] = "009001";
    ProviderIssue["XCERT_DEPLOY_GATEWAY_TAKER_NOT_EQUAL_TO_SENDER"] = "009002";
    ProviderIssue["XCERT_DEPLOY_GATEWAY_CLAIM_EXPIRED"] = "009003";
    ProviderIssue["XCERT_DEPLOY_GATEWAY_INVALID_SIGNATURE"] = "009004";
    ProviderIssue["XCERT_DEPLOY_GATEWAY_DEPLOY_CANCELED"] = "009005";
    ProviderIssue["XCERT_DEPLOY_GATEWAY_DEPLOY_ALREADY_PERFORMED"] = "009006";
    ProviderIssue["XCERT_DEPLOY_GATEWAY_MAKER_NOT_EQUAL_TO_SENDER"] = "009007";
    ProviderIssue["DAPP_TOKEN_NOT_ENOUGH_BALANCE"] = "010001";
    ProviderIssue["DAPP_TOKEN_NOT_ENOUGH_ALLOWANCE"] = "010002";
    ProviderIssue["DAPP_TOKEN_NOT_WHITELISTED_ADDRESS"] = "010003";
    ProviderIssue["DAPP_TOKEN_NOT_MIGRATION_NOT_STARTED"] = "010004";
    ProviderIssue["DAPP_TOKEN_MIGRATION_STARTED"] = "010005";
    ProviderIssue["DAPP_TOKEN_NOT_ABLE_TO_MIGRATE"] = "010006";
    ProviderIssue["DAPP_TOKEN_INVALID_SIGNATURE"] = "010007";
    ProviderIssue["DAPP_TOKEN_CLAIM_PERFORMED"] = "010008";
    ProviderIssue["DAPP_TOKEN_CLAIM_EXPIRED"] = "010009";
    ProviderIssue["DAPP_TOKEN_INVALID_SIGNATURE_KIND"] = "010010";
    ProviderIssue["DAPP_TOKEN_CLAIM_CANCELED"] = "010011";
    ProviderIssue["ERC20_DEPLOY_GATEWAY_INVALID_SIGNATURE_KIND"] = "011001";
    ProviderIssue["ERC20_DEPLOY_GATEWAY_TAKER_NOT_EQUAL_TO_SENDER"] = "011002";
    ProviderIssue["ERC20_DEPLOY_GATEWAY_CLAIM_EXPIRED"] = "011003";
    ProviderIssue["ERC20_DEPLOY_GATEWAY_INVALID_SIGNATURE"] = "011004";
    ProviderIssue["ERC20_DEPLOY_GATEWAY_DEPLOY_CANCELED"] = "011005";
    ProviderIssue["ERC20_DEPLOY_GATEWAY_DEPLOY_ALREADY_PERFORMED"] = "011006";
    ProviderIssue["ERC20_DEPLOY_GATEWAY_MAKER_NOT_EQUAL_TO_SENDER"] = "011007";
    ProviderIssue["ERC20_TRANSFER_PROXY_TRANSFER_FAILED"] = "012001";
    ProviderIssue["ACTIONS_GATEWAY_INVALID_SIGNATURE_KIND"] = "015001";
    ProviderIssue["ACTIONS_GATEWAY_INVALID_PROXY"] = "015002";
    ProviderIssue["ACTIONS_GATEWAY_SENDER_NOT_A_SIGNER"] = "015003";
    ProviderIssue["ACTIONS_GATEWAY_CLAIM_EXPIRED"] = "015004";
    ProviderIssue["ACTIONS_GATEWAY_INVALID_SIGNATURE"] = "015005";
    ProviderIssue["ACTIONS_GATEWAY_ORDER_CANCELED"] = "015006";
    ProviderIssue["ACTIONS_GATEWAY_ORDER_ALREADY_PERFORMED"] = "015007";
    ProviderIssue["ACTIONS_GATEWAY_SIGNERS_DOES_NOT_INCLUDE_SENDER"] = "015008";
    ProviderIssue["SIGNER_DOES_NOT_HAVE_ALLOW_CREATE_ASSET_ABILITY"] = "015009";
    ProviderIssue["SIGNER_DOES_NOT_HAVE_ALLOW_UPDATE_ASSET_ABILITY"] = "015010";
    ProviderIssue["SIGNER_DOES_NOT_HAVE_ALLOW_MANAGE_ABILITIES_ABILITY"] = "015011";
    ProviderIssue["SIGNER_IS_NOT_DESTROY_ASSET_OWNER"] = "015012";
    ProviderIssue["ABILITABLE_NOT_AUTHORIZED"] = "017001";
    ProviderIssue["ABILITABLE_INVALID_INPUT"] = "017002";
    ProviderIssue["OWNABLE_NOT_OWNER"] = "018001";
    ProviderIssue["OWNABLE_ZERO_ADDRESS"] = "018002";
    ProviderIssue["CLAIMABLE_NOT_PENDING_OWNER"] = "019001";
})(ProviderIssue = exports.ProviderIssue || (exports.ProviderIssue = {}));
